arangod/Aql/EngineInfoContainerDBServerServerBased.cpp

```diff
@@ -274,117 +396,128 @@ Result EngineInfoContainerDBServerServerBased::buildEngines(
-    auto res = network::sendRequest(pool, serverDest, fuerte::RestVerb::Post,
-                                    "/_api/aql/setup", std::move(buffer),
-                                    options, std::move(headers))
-                   .get();
+
+    networkCalls.emplace_back(
+        buildSetupRequest(trx, server, infoSlice, didCreateEngine, snippetIds,
+                          serverToQueryId, serverToQueryIdLock, pool, options));
+    engineInformation.emplace_back(server, infoBuilder.steal(), std::move(didCreateEngine));
     _query.incHttpRequests(unsigned(1));
-    if (res.fail()) {
-      int code = network::fuerteToArangoErrorCode(res);
-      std::string message = network::fuerteToArangoErrorMessage(res);
-      LOG_TOPIC("f9a77", DEBUG, Logger::AQL)
-          << server << " responded with " << code << " -> " << message;
-      LOG_TOPIC("41082", TRACE, Logger::AQL) << infoSlice.toJson();
-      return {code, message};
-    }
+  }
 
-    VPackSlice response = res.response->slice();
-    if (response.isNone()) {
-      return {TRI_ERROR_INTERNAL, "malformed response while building engines"};
+  futures::Future<Result> fastPathResult =
+      futures::collectAll(networkCalls)
+          .thenValue([](std::vector<arangodb::futures::Try<Result>>&& responses) -> Result {
+            // We can directly report a non TRI_ERROR_LOCK_TIMEOUT
+            // error as we need to abort after.
+            // Otherwise we need to report 
+            Result res{TRI_ERROR_NO_ERROR};
+            for (auto const& tryRes : responses) {
+              auto response = tryRes.get();
+              if (response.fail()) {
+                if (response.isNot(TRI_ERROR_LOCK_TIMEOUT)) {
+                  // Found something we cannot recover from.
+                  // Return and give up
+                  return response;
+                }
+                // track that we have lock_timeout_present
+                res = response;
+              }
+            }
+            // Return what we have, this will be ok() if and only
+            // if none of the requests failed.
+            // If will be LOCK_TIMEOUT if and only if the only error
+            // we see was LOCK_TIMEOUT.
+            return res;
+          });
+  if (fastPathResult.get().fail()) {
+    if (fastPathResult.get().isNot(TRI_ERROR_LOCK_TIMEOUT)) {
+      return fastPathResult.get();
     }
-    QueryId globalId = 0;
-    auto result = parseResponse(response, snippetIds, server, serverDest,
-                                didCreateEngine, globalId);
-    if (!result.ok()) {
-      return result;
+    {
+      // in case of fast path failure, we need to cleanup engines
+      auto requests = cleanupEngines(fastPathResult.get().errorNumber(), _query.vocbase().name(), serverToQueryId);
+      // Wait for all requests to complete.
+      // So we know that all Transactions are aborted.
+      // We do NOT care for the actual result.
+      futures::collectAll(requests).wait();
```

Comment:
Just wondering what happens if for some reason the Future(s) that we wait for here contains an error?
This will be ignore here.
Would it be better to call `get()` instead of `wait()` here and abort the entire query, because it is unsure if we were able to remove all locks that we already acquired.